### PR TITLE
iOS modular build and export implementation.

### LINKED ISSUE #33918: Automatic Godot Camera access (iOS) causes crash and App rejection
**Godot version:**
3.2 Alpha, 3.2 Beta1, 3.2 Beta2

**OS/device including version:**
Tested on iOS 12/13

**Issue description:**

Godot automatically asks for camera access at startup wich causes the app to crash on certain devices and leads to the app rejection when you try to publish it.

I could reproduce the issue on an iPhone 5S using iOS12 (works fine on iPhone X and iPad Pro and iOS13). The App does not crash when you delete the "Privacy - Camera Usage Description" but Godot still tries to get permission and the App can not be uploaded to the App store.

I am trying to release a game for children and, apart from being unable to publish it, Godot asking for Camera access on startup is quite problematic.

My code does not contain any references to the camera, microphone or arkit. Everything is deselected and I have tried to exclude them using the entitlement description.

I have tried Godot 3.1, Godot 3.2 Alpha and both Betas. I have also tried to revert to 3.1.2RC1 but it does not recognize the .tscn files after they have been opened in 3.2

At the moment it is not possible for me to release the app due to this issue.

**Steps to reproduce:**
Upload any project to iOS App store

**Minimal reproduction project:**

### PR DESCRIPTION
**Pull request by bruvzg**
Looks like Apple reject iOS apps, for simply having Camera API referenced in the code even if it's never called.

This PR implements modular iOS export, by building and exporting Godot modules as separate static libraries.

- [x] `CameraServer` init is moved form `OS` to `main` (after module type registration).
- [x] All camera access implementations are moved to the camera module.
- [x] Export template have `Camera` and `ARKit` modules as separate `libgodot_{module_name}_module.iphone.{release/debug}.{arch}.a` static libs.
- [x] iOS exporter generate stub `register/unregister_{x}_types` functions for the disabled modules.
- [x] iOS exporter extracts and adds selected static libs to the Xcode project.

Existing export template build instructions/scripts/etc. should be updated to `lipo` and archive multiple module libs.

Same approach might be useful for the other platforms and modules to disable/enable parts of the engine without fully recompiling export templates (specially for the size sensitive one like JS).

Screenshot 2019-11-29 at 15 01 18

Fixes #33918

### CODE REVIEW

File: modules/SCsub [hunk 1]
```diff
@@ -2,23 +2,30 @@
 
 Import('env')
 
-env_modules = env.Clone()
+static_build = env["platform"] == "iphone" #or env['static']
```
Comment (akien-mga) [inline modules/SCsub]: I'd prefer that moved after the `Export('env_modules')` for consistency with other `SCsub`, where import, clone and export are always the first operations.

File: modules/SCsub [hunk 2]
```diff
@@ -2,23 +2,30 @@
...
 
+env_modules = env.Clone()
 Export('env_modules')
 
-env.modules_sources = []
-
-env_modules.add_source_files(env.modules_sources, "register_module_types.gen.cpp")
+if (not static_build):
```
Comment (akien-mga) [inline modules/SCsub]: No need for parentheses around the main condition in python.

File: modules/SCsub (hunk 1 above)
Comment (akien-mga) [reply modules/SCsub]: The `env['static']` part was used for testing I guess?

BTW I wonder if `static` is a good name for that. I'm not sure what would be a good name, but we do link everything statically in the `not static_build` case. Maybe `modular_libs` or similar?

**Comment by akien-mga** [comment]
>  Same approach might be useful for the other platforms to disable/enable parts of the engine without fully recompiling export templates (specially for the size sensitive one like JS).

That's my main "concern" with this PR in the scope of 3.2 (for which we need a fix to #33918).
The approached used here to make all modules opt-out sounds interesting, and I can see many users interested in it also for other platforms as mentioned above. But that sounds like something that we should take more time to design for all platforms that can support it and ship with 4.0. And thus for 3.2 I'd prefer a more minimal solution that doesn't give users too much flexibility on iOS that will be missing on other platforms.

Can we only have the `camera` (and if needed `microphone`) modules using this opt-out approach for 3.2, and see how to expand this later on?

**Comment by bruvzg** [comment]
> Can we only have the camera (and if needed microphone) modules using this opt-out approach for 3.2, and see how to expand this later on?

Well, fully working and convenient implementation will need much more work anyway.

I'll change it to only use separate `Camera` and `ARKit` libs, AFAIK there were no complaints about microphone permission, so it's probably OK to leave it in main lib.

**Comment by akien-mga** [comment]
@pip-byte @oeleo1 @HEAVYPOLY Would one of you be able to build and test this PR? (See https://docs.godotengine.org/en/latest/development/compiling/compiling_for_ios.html)

@bruvzg could only test with the simulator as he lacks a mobile Apple device to test with. It would also be important to try the publishing process to see if the Apple Store is happy with the presence or absence of camera code based on how you exported the project.

**Comment by bruvzg** [comment]
When testing it make sure to rebuild both iOS export template and macOS editor used to export it, since changes affect export process as well. And also to `lipo` and add to the export template zip additional libraries:

For example for release x86-64 (simulator) + arm64 (device) build:
```sh
lipo -create bin/libgodot_arkit_module.iphone.opt.arm64.a bin/libgodot_arkit_module.iphone.opt.x86_64.a -output /root/out/libgodot_arkit_module.iphone.opt.fat
lipo -create bin/libgodot_camera_module.iphone.opt.arm64.a bin/libgodot_camera_module.iphone.opt.x86_64.a -output /root/out/libgodot_camera_module.iphone.opt.fat
```

**Comment by bruvzg** [comment]
Here is my build to test (includes macOS `.app` and `.tpz` with iOS export template):

https://github.com/bruvzg/godot/releases/tag/v3.2_ios_test

**Comment by oeleo1** [comment]
@bruvzg quick testing your build & templates gave the following results:

 * export with ArKit = no, Camera = no: app launches as expected, no camera access screen asked, all is well. Congrats.
 * export with Arkit = no, Camera = yes: app launches, camera access popup as expected. Chosing Deny, works as expected, app proceeds. Choosing Allow crashes the app (Invalid thread access)
 * export with Arkit = yes, Camera = yes or no: Xcode link error
`"ClassDB::bind_methodfi(unsigned int, MethodBind*, char const*, Variant const**, int)", referenced from: ARKitInterface::_bind_methods() in libgodot_arkit_module.a(arkit_interface.iphone.opt.arm64.o)`

Next, I'll try to build your sources myself. I am using the attached build.sh which I adapted from your instructions - could you please check & fix it cf. what needs to be done here so that I can give it a shot. Thanks.
[build.sh.txt](https://github.com/godotengine/godot/files/3908465/build.sh.txt)

**Comment by bruvzg** [comment]
>  could you please check & fix it

```
cp bin/libgodot_module_arkit.iphone.debug.universal.a misc/dist/ios_xcode/libgodot_module_arkit.iphone.release.fat.a
cp bin/libgodot_module_camera.iphone.debug.universal.a misc/dist/ios_xcode/libgodot_module_camera.iphone.release.fat.a
```

On lines 78-79 debug libs are copied to the `release.fat.a`, otherwise everything seems OK

> export with Arkit = yes, Camera = yes or no: Xcode link error ClassDB::bind_methodfi

Oops, I missed debug/release check for the new libs and they got mixed, should be fixed by last push (only editor need to be recompiled, templates should be OK as is), same issue might also cause crash.

**Comment by bruvzg** [comment]
> Choosing Allow crashes the app (Invalid thread access)

I was pretty sure app with camera usage should crash on the simulator, but I checked camera code once more and found another problem, [NSArray arrayWithObjects](https://developer.apple.com/documentation/foundation/nsarray/1460145-arraywithobjects) init list should end with the NIL.

After adding it, app in the simulator gives camera access prompt and don't crash any more.

**Comment by oeleo1** [comment]
@bruvzg Okay, I've built and tested your latest branch -- it looks good for all export configurations. Bravo!

 * ArKit = no, Camera = no: no camera prompt, app runs
 * ArKit = no, Camera = yes: camera prompt, Deny => app runs, Allow => app runs, no crash
 * ArKit = yes, Camera = no: no camera prompt (?), app runs
 * ArKit = yes, Camera = yes: camera prompt, Allow/Deny => app runs, no crash

I haven't tested the app publishing though -- never did that. You may want to update your test branch for @pip-byte to download the build & templates to go over it.

**Comment by akien-mga** [comment]
Awesome work @bruvzg and thanks for the thorough testing @oeleo1!

Let's merge as is, but we'll indeed need a confirmation that it's working as expected on the App Store (either from a test build or the upcoming 3.2 beta 3).

**Comment by bruvzg** [comment]
> You may want to update your test branch for @pip-byte to download the build & templates to go over it.

Test build is already updated (same link in previous comment).

> ArKit = yes, Camera = no: no camera prompt (?)

It's probably OK, ARKit is not initialized, unless you create session from the script, but it potentially might trigger same App-Store issues.

**Comment by bruvzg** [comment]
We really should fix all the [iOS build warnings](#29800), crash inducing issue was always reported there, but it's hard to catch anything meaningful in the tones of deprecation and mixed sign comparison stuff.

```
platform/iphone/camera_ios.mm:362:301: warning: missing sentinel in method dispatch [-Wsentinel]
        AVCaptureDeviceDiscoverySession *session = [AVCaptureDeviceDiscoverySession discoverySessionWithDeviceTypes:[NSArray arrayWithObjects:AVCaptureDeviceTypeBuiltInTelephotoCamera, AVCaptureDeviceTypeBuiltInDualCamera, AVCaptureDeviceTypeBuiltInTrueDepthCamera, AVCaptureDeviceTypeBuiltInWideAngleCamera] mediaType:AVMediaTypeVideo position:AVCaptureDevicePositionUnspecified];
note: method has been explicitly marked sentinel here
arrayWithObjects:(ObjectType)firstObj, ... NS_REQUIRES_NIL_TERMINATION;
```

**Comment by HEAVYPOLY** [comment]
I'm probably doing something wrong since I've never compiled for linux before, but when trying to open the exported project in xcode I'm getting this error  
image

**Comment by akien-mga** [comment]
> We really should fix all the iOS build warnings (#29800)

There's a wealth of fixes in #22784, but it was never brought to a mergeable state sadly. Would definitely be worth cherry-picking/squashing relevant changes there to fix build warnings.

**Comment by HEAVYPOLY** [comment]
Maybe related, getting this error when trying to build linux export template from mac through docker on 3.x
```[ 96%] Linking Program        ==> bin/godot.x11.opt.64
[100%] progress_finish(["progress_finish"], [])
[100%] /usr/bin/ld: modules/libmodules.x11.opt.64.a(register_module_types.gen.x11.opt.64.o): in function `register_module_types()':
register_module_types.gen.cpp:(.text+0xa): undefined reference to `register_camera_types()'
/usr/bin/ld: modules/libmodules.x11.opt.64.a(register_module_types.gen.x11.opt.64.o): in function `unregister_module_types()':
register_module_types.gen.cpp:(.text+0xba): undefined reference to `unregister_camera_types()'
collect2: error: ld returned 1 exit status
scons: *** [bin/godot.x11.opt.64] Error 1
scons: building terminated because of errors.```

**Pull request merged**